sfa_dash/blueprints/form.py

```diff
@@ -1,42 +1,277 @@
-        json_dict = self.request_schema()
-        form_fields = {}
-        for key, value in json_dict[0]['definitions']['SiteSchema']['properties'].items():
-            form_fields[key] = {
-                'type': self.type_map.get(value['type'], None),
-                'label': value.get('title', key),
-                'name': key,
-                'description': value.get('description', None ),
-                'required': key in json_dict[0]['definitions']['SiteSchema']['required'],
-            }
-        form = render_template('forms/create_form.html', form_fields=form_fields)
-        return form
+    def flatten_dict(self, to_flatten):
+        """Flattens nested dictionaries, removing keys of the nested elements.
+        Useful for flattening API responses for prefilling forms on the
+        dashboard.
+        """
+        flattened = {}
+        for key, value in to_flatten.items():
+            if isinstance(value, dict):
+                flattened.update(self.flatten_dict(value))
+            else:
+                flattened[key] = value
+        return flattened
+
+    def parse_hhmm_field(self, data_dict, key_root):
+        """ Extracts and parses the hours and minutes inputs to create a
+        parseable time of day string in HH:MM format. These times are
+        displayed as two select fields designated with a name (key_root)
+        and _hours or _minutes suffix.
+
+        Parameters
+        ----------
+        data_dict: dict
+            Dictionary of posted form data
+
+        key_root: string
+            The shared part of the name attribute of the inputs to parse.
+            e.g. 'issue_time' will parse and concatenate 'issue_time_hours'
+            and 'issue_time_minutes'
+
+        Returns
+        -------
+        string
+            The time value in HH:MM format.
+        """
+        hours = int(data_dict[f'{key_root}_hours'])
+        minutes = int(data_dict[f'{key_root}_minutes'])
+        return f'{hours:02d}:{minutes:02d}'
+
+    def parse_timedelta(self, data_dict, key_root):
+        value = int(data_dict[f'{key_root}_number'])
+        units = data_dict[f'{key_root}_units']
+        if units == 'minutes':
+            return value
+        if units == 'hours':
+            return value * 60
+        else:
+            return value * 3600
+
+    def site_formatter(self, site_dict):
+        """Formats the result of a site webform into an API payload.
 
+        Parameters
+        ----------
+        site_dict:  dict
+            The posted form data parsed into a dict.
+        Returns
+        -------
+        dictionary
+            Form data formatted to the API spec.
+        """
+        modeling_keys = ['ac_capacity', 'dc_capacity',
+                         'temperature_coefficient', 'axis_azimuth',
+                         'tracking_type', 'backtrack',
+                         'axis_tilt', 'ground_coverage_ratio',
+                         'surface_tilt', 'surface_azimuth']
```

Comment:
we added `dc_losses` and `ac_losses` to the data model.

Reply:
:+1: I'll add this to the API as well.